packages/eslint-plugin/src/rules/max-params.ts

```diff
@@ -0,0 +1,99 @@
+import { AST_NODE_TYPES, type TSESTree } from '@typescript-eslint/utils';
+import type { JSONSchema4 } from '@typescript-eslint/utils/json-schema';
+
+import * as util from '../util';
+import { getESLintCoreRule } from '../util/getESLintCoreRule';
+
+type FunctionLike =
+  | TSESTree.FunctionDeclaration
+  | TSESTree.FunctionExpression
+  | TSESTree.ArrowFunctionExpression;
+
+type FunctionRuleListener<T extends FunctionLike> = (node: T) => void;
+
+const baseRule = getESLintCoreRule('max-params');
+
+export type Options = util.InferOptionsTypeFromRule<typeof baseRule>;
+export type MessageIds = util.InferMessageIdsTypeFromRule<typeof baseRule>;
+
+const schema = Object.values(
+  util.deepMerge(
+    { ...baseRule.meta.schema },
+    {
+      0: {
+        oneOf: [
+          baseRule.meta.schema[0].oneOf[0],
+          {
+            ...baseRule.meta.schema[0].oneOf[1],
+            properties: {
+              ...baseRule.meta.schema[0].oneOf[1].properties,
+              countVoidThis: {
+                type: 'boolean',
+              },
+            },
+          },
+        ],
+      },
+    },
+  ),
+) as JSONSchema4[];
```

Comment:
for some reason, I can't get `deepMerge` to work properly here. 

I've tried adding these lines to add support for array overrides, but I'm not sure why it doesn't work:
```diff
if (isObjectNotArray(firstValue) && isObjectNotArray(secondValue)) {
  // object type
  acc[key] = deepMerge(firstValue, secondValue);
+ } else if (Array.isArray(firstValue)) {
+   // array type
+   acc[key] = Object.values(
+     deepMerge({ ...firstValue }, { ...(secondValue as ObjectLike) }),
+   );
+ } else {
  // value type
  acc[key] = secondValue;
}
```

seems to work fine when isolated though:
https://tsplay.dev/wX1GLN